metaflow/client/cache/cache_server.py

```diff
@@ -0,0 +1,307 @@
+        if keys:
+            fields.extend(sorted(keys))
+        if stream_key:
+            fields.append(stream_key)
+        token = hashlib.sha1('|'.join(fields).encode('utf-8')).hexdigest()
+    else:
+        token = idempotency_token
+
+    return {
+        'op': op,
+        'action': action,
+        'priority': prio,
+        'keys': keys,
+        'stream_key': stream_key,
+        'message': message,
+        'idempotency_token': token,
+        'disposable_keys': disposable_keys
+    }
+
+def echo(msg):
+    now = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%S.%f')
+    sys.stderr.write('CACHE [%s] %s\n' % (now, msg))
+
+class MessageReader(object):
+
+    def __init__(self, fd):
+        # make fd non-blocking
+        fl = fcntl.fcntl(fd, fcntl.F_GETFL)
+        fcntl.fcntl(fd, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+        self.buf = io.BytesIO()
+        self.fd = fd
+
+    def messages(self):
+        while True:
+            try:
+                b = os.read(self.fd, 65536)
+                if not b:
+                    return
+            except OSError as e:
+                if e.errno == 11: # EAGAIN
+                    return
+            else:
+                self.buf.write(b)
+                if b'\n' in b:
+                    new_buf = io.BytesIO()
+                    self.buf.seek(0)
+                    for line in self.buf:
+                        if line.endswith(b'\n'):
+                            try:
+                                yield json.loads(line)
+                            except:
+                                raise
+                                uni = line.decode('utf-8', errors='replace')
+                                echo("WARNING: Corrupted message: %s" % uni)
+                        else:
+                            # return the last partial line back to the buffer
+                            new_buf.write(line)
+                    self.buf = new_buf
+
+    def close(self):
+        tail = self.buf.getvalue()
+        if tail:
+            uni = tail.decode('utf-8', errors='replace')
+            echo("WARNING: Truncated message: %s" % uni)
+
+def subprocess_cmd_and_env(mod):
+    pypath = os.environ.get('PYTHONPATH', '')
+    env = os.environ.copy()
+    env['PYTHONPATH'] = ':'.join((os.getcwd(), pypath))
+    return [sys.executable, '-m', 'metaflow.client.cache.%s' % mod], env
+
+class Worker(object):
+
+    def __init__(self, request, filestore):
+        self.request = request
+        self.prio = request['priority']
+        self.filestore = filestore
+        self.proc = None
+        self.tempdir = self.filestore.open_tempdir(request['idempotency_token'],
+                                                   request['action'],
+                                                   request['stream_key'])
+        if self.tempdir is None:
+            self.echo("Store couldn't create a temp directory. "\
+                      "WORKER NOT STARTED.")
```

Comment:
Referring to my other comment about worker functionality depending on a tempdir; shouldn't this case be an explicit exception?